spring-cloud-azure-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java

```diff
@@ -5,27 +5,50 @@
  */
 package com.microsoft.azure.spring.cloud.config;
 
-import com.microsoft.azure.spring.cloud.autoconfigure.telemetry.TelemetryAutoConfiguration;
 import com.microsoft.azure.spring.cloud.autoconfigure.telemetry.TelemetryCollector;
+import com.microsoft.azure.spring.cloud.config.msi.AzureConfigMSIConnector;
+import com.microsoft.azure.spring.cloud.config.msi.ConfigAccessKeyResource;
+import com.microsoft.azure.spring.cloud.config.msi.ConfigMSICredentials;
 import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.impl.client.HttpClients;
-import org.springframework.boot.autoconfigure.AutoConfigureBefore;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
 
 import javax.annotation.PostConstruct;
 
 @Configuration
-@AutoConfigureBefore(TelemetryAutoConfiguration.class)
 @EnableConfigurationProperties(AzureCloudConfigProperties.class)
 @ConditionalOnClass(AzureConfigPropertySourceLocator.class)
 @ConditionalOnProperty(prefix = AzureCloudConfigProperties.CONFIG_PREFIX, name = "enabled", matchIfMissing = true)
 public class AzureConfigBootstrapConfiguration {
     private static final String AZURE_CONFIG_STORE = "AzureConfigService";
 
+    @Bean
+    @Primary
+    public AzureCloudConfigProperties initProperties(AzureCloudConfigProperties properties) {
+        if (StringUtils.hasText(properties.getConnectionString())) {
+            return properties;
+        }
+
+        ConfigMSICredentials msiCredentials = new ConfigMSICredentials(properties.getMsi());
+        ConfigAccessKeyResource keyResource = new ConfigAccessKeyResource(properties.getMsi());
+
+        AzureConfigMSIConnector msiConnector = new AzureConfigMSIConnector(msiCredentials, keyResource);
+        properties.setConnectionString(msiConnector.getConnection());
+        properties.validateAndInit();
```

Comment:
Since validateAndInit only needed when connection string changed, I suggest to make it inside setter of connection string.

Reply:
validateAndInit  is a `PostConstruct` method, has to be kept in Properties class to ensure property configuration. But this reminds me the `properties.validateAndInit();` is not necessary, `ConfigurationPropertiesBindingPostProcessor` will ensure it's called.